### ISSUE TITLE
Use Gradle's task configuration avoidance API

### ISSUE THREAD

**Issue by rymanske** [ImFlog/schema-registry-plugin#34]
I'd like to propose migrating to Gradle's task configuration avoidance as outlined [here](https://docs.gradle.org/5.6.4/userguide/task_configuration_avoidance.html).

I'm trying to check compatibility of schemas that have been generated from Avro IDL files using [davidmc24/gradle-avro-plugin](https://github.com/commercehub-oss/gradle-avro-plugin).

Currently all plugin configuration in the `schemaRegistry` block is executed immediately when running any Gradle task. Ideally this configuration would execute only when the task that requires the config executes.

The following does not work, as configuration happens before schema files are generated:
```groovy
schemaRegistry {
    url = "http://schema-registry.example.com:8081"
    compatibility {
        fileTree(dir: "build/generated-avro-avsc", include: "**/*.avsc").each { file ->
            subject("subjectName", relativePath(file))
        }
    }
}
```

Instead I need to do this:
```groovy
schemaRegistry {
    url = "http://schema-registry.example.com:8081"
}

tasks.whenTaskAdded { task ->
    if (task.name == "testSchemasTask") {
        task.doFirst {
            schemaRegistry {
                compatibility {
                    fileTree(dir: "build/generated-avro-avsc", include: "**/*.avsc").each { file ->
                        subject("subjectName", relativePath(file))
                    }
                }
            }
        }
    }
}
```

**Comment by ImFlog** [comment]
Hello, 

Thank you for your interest in this plugin. I didn't knew about this feature (probably wasn't here when I first created the plugin).
This sound interesting and I will try to migrate the plugin to this new API (probably start to hack this during the weekend) but if you are willing to, PR are welcome :)

Also, I am very interested about the gradle-avro-plugin as it is something i wanted to provide inside the plugin directly. Is your work available as an open source project ?

**Comment by rymanske** [comment]
Thanks for your response!

What I've been working on isn't open source, but I put together [this example](https://github.com/rymanske/schema-registry-plugin-compatibility-sample) that demonstrates how I'm using the two plugins.

I'd be happy to contribute when I have time this week.

**Comment by rymanske** [comment]
I played around with this a bit and I think I was misunderstanding the way this Gradle feature worked. It won't stop things like the creation of a `fileTree` outside of the plugin from executing immediately. I 

To do what I want, something like this is more appropriate:
```groovy
task configureSchemaRegistryPlugin {
    dependsOn tasks.generateSchema

    schemaRegistry {
        url = "http://schema-registry-rul.com:8081"
    }

    doLast {
        schemaRegistry {
            compatibility {
                fileTree(dir: "build/generated-avro-avsc", include: "**/*.avsc").each { file ->
                    subject("subjectName", relativePath(file))
                }
            }
        }
    }
}

tasks.whenTaskAdded { task ->
    if (task.name == "testSchemasTask") {
        task.dependsOn tasks.configureSchemaRegistryPlugin
    }
}
```

 [This documentation](https://docs.gradle.org/5.6.4/userguide/lazy_configuration.html) on lazy configuration was helpful.

What do you think about exposing task types that can be configured manually? [Here's an example](https://github.com/davidmc24/gradle-avro-plugin/blob/master/README.md#alternate-usage) from the gradle-avro-plugin.

**Comment by ImFlog** [comment]
To be honest I've not yet had time to take a deeper look, but I will !
If in the meantime you want to hack a PR that shows what you want to do It would be great !

**Comment by ImFlog** [comment]
Ok I finally found time to work on this issue, I use the register method and lazy properties so you should be be able to use avro outputs as input in the tasks.
Can you try to build the plugin locally and see if it fits ?
If so I will release this new version soon. It's a pretty big refactoring but I'm happy for with the result :)

**Comment by ImFlog** [comment]
@rymanske, I've just released the version 0.9.0. Can you check if your use case fits ?
Also I'm planning to add an example directory, if your use case work could you submit a PR with the schema registry + avro plugin ?

**Comment by rymanske** [comment]
Hi, sorry for the delay and thanks for your help. I made [an update](https://github.com/rymanske/schema-registry-plugin-compatibility-sample/commit/c0a90360922ba86c15111180defa22716e85dc4e) to my example code so that it works with the new version. I think I originally misunderstood what lazy configuration could do, so now that I understand how it works, I think this way of configuring the plugin is acceptable. 

I think my use case might be fairly niche, but if you think otherwise I can create a PR with my example. What do you have in mind for the example directory?

**Comment by ImFlog** [comment]
I don't understand why you need to create a new task with the doFirst part in your example.
Can you elaborate on this ?
Anyway I want to add basic example like generating avro classes from the registry using this plugin and the avro one.

I close this task if the current implementation suits you. answer there anyway so I can understand better :)
Thank you for you input!

**Comment by rymanske** [comment]
I found that when configuring a task, everything outside of a `doFirst` or `doLast`is evaluated before any tasks are executed. So for my example this means that if I run the `build` task, without the `doFirst`, the `schemaRegistry` closure is evaluated and we look for Avro schemas in the build dir. After that, the `build` task will run. The `testSchemasTask` won't do anything, since the build dir was empty when it was configured.

One thing I ran into when testing locally is that the `testSchemasTask` will be evaluated before the `clean` task is executed, so it will find schemas left from the last build if the `clean` task is not run separately.

So when testing I run
```bash
./gradlew clean
```
and then 
```bash
./gradlew build
```


Let me know if that makes sense.
